agent.py

```diff
@@ -98,26 +101,82 @@ def _handle_sighup(self, signum, frame):
         log.info("Attempting a configuration reload...")
+        hostname = get_hostname(self._agentConfig)
 
-        # Stop checks
-        for check in self._checksd.get('initialized_checks', []):
-            check.stop()
+        # if no check was given, reload them all
+        if not checks_to_reload:
+            log.debug("No check list was passed, reloading every check")
+            # stop checks
+            for check in self._checksd.get('initialized_checks', []):
+                check.stop()
 
-        # Reload checksd configs
-        hostname = get_hostname(self._agentConfig)
-        self._checksd = load_check_directory(self._agentConfig, hostname)
+            self._checksd = load_check_directory(self._agentConfig, hostname)
+        else:
+            new_checksd = copy(self._checksd)
+
+            self.refresh_specific_checks(hostname, new_checksd, checks_to_reload)
+            # once the reload is done, replace existing checks with the new ones
+            self._checksd = new_checksd
 
         # Logging
         num_checks = len(self._checksd['initialized_checks'])
         if num_checks > 0:
-            log.info("Successfully reloaded {num_checks} checks".
-                     format(num_checks=num_checks))
+            opt_msg = " (refreshed %s checks)" % len(checks_to_reload) if checks_to_reload else ''
+
+            msg = "Check reload was successful. Running {num_checks} checks{opt_msg}.".format(
+                num_checks=num_checks, opt_msg=opt_msg)
+            log.info(msg)
         else:
             log.info("No checksd configs found")
 
+    def refresh_specific_checks(self, hostname, checksd, checks):
+        """take a list of checks and for each of them:
+            - remove it from the init_failed_checks if it was there
+            - load a fresh config for it
+            - replace its old config with the new one in initialized_checks if there was one
+            - disable the check if no new config was found
+            - otherwise, append it to initialized_checks
+        """
+        for check_name in checks:
+            idx = None
+            for num, check in enumerate(checksd['initialized_checks']):
+                if check.name == check_name:
+                    idx = num
+                    # stop the existing check before reloading it
+                    check.stop()
+
+            if not idx and check_name in checksd['init_failed_checks']:
+                # if the check previously failed to load, pop it from init_failed_checks
+                checksd['init_failed_checks'].pop(check_name)
+
+            fresh_check = load_check(self._agentConfig, hostname, check_name)
+
+            # this is an error dict
+            # checks that failed to load are added to init_failed_checks
+            if isinstance(fresh_check, dict) and 'error' in fresh_check.keys():
+                checksd['init_failed_checks'][fresh_check.keys()[0]] = fresh_check.values()[0]
+
```

Comment:
here, if the check failed to initialize but was previously loaded, don't we want to remove it from `checksd['initialized_checks']` too?